### PR TITLE
[#1004] Prove the pure logic with property-based tests rather than with examples

### LINKED ISSUE #1004: Prove the pure logic with property-based tests rather than with examples
## Context

Every one of the suite's tests is an example: it fixes an input and asserts the output. The defects that reached production were not on uncovered paths but on inputs nobody thought of — a base64url value whose final character carries unused bits, a recurrence crossing a DST boundary, a redaction pass that hid one shape of a value and not another. A property states what must hold for *every* input in a domain and lets a generator hunt for the counterexample, which is the one thing an example cannot do. Nothing in `Directory.Packages.props` provides that today.

The targets are the pure, deterministic logic where an invariant can be written down at all — encode/decode pairs, coverage of a source by its parts, ordering, and the redaction guarantee this project already states as *prefer over-hiding*.

## User stories

**As the** maintainer **I want** the pure logic checked against generated inputs, **so that** an edge case is found by a test run rather than by the live deployment.

**As a** reviewer **I want** an invariant stated once as a property, **so that** an implementation change is judged against the rule rather than against whichever examples somebody happened to write.

**As a** contributor **I want** a failing property to report a shrunk counterexample and its seed, **so that** the bug arrives as a two-line repro rather than as a search.

## Acceptance

- A property-based testing library is chosen, its licence reviewed and recorded in `THIRD_PARTY_LICENSES.md`, its version pinned in `Directory.Packages.props`, and the lock files regenerated in the same change.
- The library runs under xUnit v3 on Microsoft Testing Platform v2 and shrinks a failing case before reporting it.
- Generation is seeded and the failing seed is reported, so a counterexample is reproducible rather than observed once.
- Properties cover at least: the keyset pagination cursor's encode/decode roundtrip, chunking (no source text lost, offsets monotonic, passages covering the source), the next-occurrence calculation across a DST boundary, and redaction (a redacted value never appears in the output, for any generated input).
- Each property names its invariant in its own test name, so a failure reads as the rule that broke.
- The iteration count is a constant per class rather than a literal per test, and the whole set runs within the ordinary unit-test budget.
- `tests/AGENTS.md` states when a property is written and when an example still is.

## Out of scope

Replacing existing example tests. A property covers a rule and an example covers a case that once broke; both stay.

Properties over anything that reaches the database, an IMAP session, or a provider. Generation multiplies the cost of an impure test until it is no longer runnable.

## Dependencies

Informed by the mutation-score report from #1002 — surviving mutants say which types carry assertions that do not hold — but not blocked by it.

### PR DESCRIPTION
**Pull request by Krzysztof318**
Every test in the suite was an example: an input fixed, an output asserted. The defects that reached production here were on inputs nobody thought of rather than on uncovered paths, and an example cannot hunt for those. This adds the generator that can, and states ten properties over the pure logic where an invariant can be written down.

## The library

**CsCheck 4.8.0**, Apache-2.0, test-only. It was chosen over the alternatives on three things the issue asks for directly:

- **It couples to no test framework.** A property is an ordinary `[Fact]` that calls into it, so nothing about xUnit v3 discovery, Microsoft Testing Platform v2, the runner, or the reporting changes, and there is no framework-integration package whose xUnit v3 support has to be checked at every upgrade.
- **It carries no dependency of its own** on its `net8.0` target — the three regenerated lock files show one added row each and no transitive one.
- **Shrinking and the seed are its design rather than a feature.** Generation and shrinking are both driven by PCG, so a failure reports the shrunk input, the reason, and a seed that replays exactly that case through `CsCheck_Seed`.

Pinned in `Directory.Packages.props`, recorded in `THIRD_PARTY_LICENSES.md` under *Test-only*, and `tests/Application.UnitTests/packages.lock.json`, `tests/AI.UnitTests/packages.lock.json`, and `tests/SharedSources.UnitTests/packages.lock.json` are regenerated in the same change.

## What is stated

`PropertyCheck` in `tests/shared` is the one way a property is sampled. It fixes the thread count — the suite already runs its collections in parallel, so a property fanning out across every logical processor would change the timing of every test beside it — and it is where the reproduction contract is documented once instead of per suite. `PropertyCheckTests` asserts that contract rather than assuming it: that a counterexample is shrunk to the simplest input that breaks the invariant, and that the seed named beside it replays the same case.

| Subject | Properties |
|---|---|
| `KeysetCursorPayload` | Every payload a reading can issue survives the encoding; any text a caller presents is either refused or decodes to a payload that encodes back to itself |
| `DeterministicEmailTextChunker` | Every passage reads back from its own offsets; the passages are numbered by a strictly advancing offset; no non-blank character of the text falls outside every passage — for any text and any rule set |
| `JobRecurrence` | The next occasion is the earliest one strictly after the instant; the backward reading answers that occasion itself; the counted window agrees with walking it |
| `SensitiveContentRedactor` | No detected value is readable in the result; exactly the characters no finding covered survive |

The recurrence properties are sampled around **every clock change four zones make** over three years, discovered by scanning the zone's own offset rather than written down: `Europe/Warsaw`, `America/New_York`, `Australia/Sydney` — two hemispheres, so the change falls in opposite seasons — and `Australia/Lord_Howe`, which shifts by half an hour rather than a whole one. A `[Fact]` beside them asserts that corpus is real, so a machine whose time-zone data named no change fails loudly instead of leaving four properties sampling schedules that never move.

## Each property was checked against a deliberate break

A property over a generator that never produces the interesting shape passes forever and reads exactly like one that works. So each was run against a mutation of the production code and put back:

| Mutation | Caught by |
|---|---|
| The cursor writes the position in its own offset rather than as UTC ticks | the round trip |
| The next chunk window starts one character past the previous end | the coverage property |
| The next occasion may land on the instant itself | the earliest-occasion property |
| Overlapping findings are dropped instead of covered | both redaction properties |
| The analyzed ceiling emits the unscanned remainder as well | both redaction properties |

The last two are why the redaction generator draws two detected values, the second often built to start inside the first: the first version of these properties drew one value, produced no overlap, and passed the merge mutation. The second redaction property exists for the same reason — *a value is unreadable* is still satisfied by a region covering only part of an overlap, and *exactly the characters no finding covered survive* is not.

## Documentation

`tests/AGENTS.md` gains a *Properties and examples* section: when a property is written, when an example still is, that neither replaces the other, that nothing impure is generated over, and the non-vacuity check above as a rule rather than as something this change happened to do. It also carries the one rule a property relaxes — the unit-test policy forbids nondeterministic values, and a property draws different inputs on each run — with what pays for it: a property must be true of every input its generator can produce, so a red property is a defect and never a re-run.

`docs/architecture/solution-structure.md` § *Test projects* named xUnit v3 and NSubstitute as the suite's tooling and had stopped being complete.

## Notes for review

- **No public contract moves.** No MCP tool descriptor, configuration key, database column, or deployment asset changes, so `CHANGELOG.md` is untouched and the golden files from #1069 are unchanged.
- **Nothing under `src/` changes.** The mutations above were reverted; `git status` after the full gate is clean.
- **Residual risk, stated rather than hidden:** a property samples different inputs on each run, so one that is only usually true would surface on somebody else's pull request. Each of the ten is true of every input its generator can produce by construction, and the mutation table is the evidence that they are also not vacuous. The recurrence corpus assertion depends on the host's time-zone data still giving those four zones two changes a year; if a zone abolishes daylight saving, that `[Fact]` fails by design rather than the properties quietly going empty.
- The whole set costs about two seconds of the unit run: 10409 tests pass, aggregate line coverage 95.70%.

Closes #1004

### CODE REVIEW

File: THIRD_PARTY_LICENSES.md [hunk 1]
```diff
@@ -118,6 +118,7 @@ Referenced by the unit and integration suites and by nothing that ships.
...
+| CsCheck 4.8.0 | The generator behind the property tests, which state an invariant over generated inputs instead of over a chosen one: the shared keyset cursor codec, the chunking boundaries, the daily recurrence around a clock change, and the redaction guarantee. Referenced by `tests/Application.UnitTests`, `tests/AI.UnitTests`, and `tests/SharedSources.UnitTests`, and reached only through `PropertyCheck` in `tests/shared`. | Apache-2.0, declared as a license expression in the NuGet package metadata and confirmed against the upstream repository | Allowed for test projects. Not redistributed, so no notice travels with an artifact. It carries no dependency of its own on its `net8.0` target and couples to no test framework, so it adds one assembly to three unit-test projects and changes nothing about how the suite is discovered or run. | <https://github.com/AnthonyLloyd/CsCheck>, <https://www.nuget.org/packages/CsCheck/4.8.0> |
```
[the remark this answers is not in the document]
Comment (Krzysztof318) [reply THIRD_PARTY_LICENSES.md]: Correct on both counts — fixed in 2f9d0e0a.

The row overstated it. Each `*Properties.cs` file does build its generators against the `Gen` API directly, with its own `using CsCheck;`; what goes only through `PropertyCheck` is the sampling, which is the part that had to be one decision (thread count and the reproduction contract) rather than one per suite. `PropertyCheckTests` is the one file that calls `Sample` itself, because the contract it asserts is that sampler's own — and that is the exception `tests/AGENTS.md` already documents.

`THIRD_PARTY_LICENSES.md:121` now says exactly that, and names the exception; `docs/architecture/solution-structure.md:129` carries the same correction, since it made the identical claim.

**Comment by Krzysztof318** [comment]
Two items from the review body, neither of which has a thread to close.

**The licence.** The review lists MIT as a possible reading of CsCheck 4.8.0. It is **Apache-2.0**, from the artifact rather than from a listing: the `.nuspec` inside `cscheck.4.8.0.nupkg` carries `<license type="expression">Apache-2.0</license>`, and `gh api repos/AnthonyLloyd/CsCheck/license` returns spdx_id `Apache-2.0` for the repository the package points at. That is what `THIRD_PARTY_LICENSES.md` records, so no row changes.

**Where CsCheck is reached from.** That one had a thread and is fixed in 2f9d0e0a — both `THIRD_PARTY_LICENSES.md` and `docs/architecture/solution-structure.md` now say that each property class builds its generators against the `Gen` API directly and samples them through `PropertyCheck`, with `PropertyCheckTests` named as the one exception.

Head is now 2f9d0e0a; `scripts/verify-full.sh` is green on it (10409 tests, aggregate line coverage 95.70%, contract suite 238 passed).
